EventSourcing.Cosmos/StoredProcedures/DeleteAggregateAll.js

```diff
@@ -0,0 +1,117 @@
+ *
+ * @function
+ * @param {string} containerId - The id of the container where the aggregate exists
+ * @param {string} partitionId - The partition id of the aggregate that is to be deleted
+ * @param {string} aggregateId - The aggregate id of the aggregate that is to be deleted
+ * @returns {Object.<number, boolean>} Returns an object with the two properties:<br/>
+ *   deleted - contains a count of documents deleted<br/>
+ *   continuation - a boolean whether you should execute the sproc again (true if there are more documents to delete; false otherwise).
+ */
+function deleteAggregateAll(containerId, partitionId, aggregateId) {
+    var collection = getContext().getCollection();
+    var collectionLink = collection.getSelfLink();
+    var response = getContext().getResponse();
+    var responseBody = {
+        deleted: 0,
+        continuation: true
+    };
+
+    // Validate input.
+    if(!containerId) throw new Error('Invalid container id');
+    if(!partitionId) throw new Error('Invalid partition id');
+    if(!aggregateId) throw new Error('Invalid aggregate id');
+    
+    var query = `SELECT * FROM ${containerId} e WHERE e.PartitionId = '${partitionId}' AND e.AggregateId = '${aggregateId}'`;
+    
+    // Get the current aggregate version -> Create reservation event -> Delete reservation event -> Query and delete all documents related to the aggregate
+    getIndex();
+
+    // Find the current version of the aggregate by getting the maximal Index of the events
+    function getIndex() {
+        var versionQuery = `SELECT Max(e.Index) AS Index FROM ${containerId} e WHERE e.PartitionId = '${partitionId}' AND e.AggregateId = '${aggregateId}'`;
+        var isAccepted = collection.queryDocuments(collectionLink, versionQuery, {}, createReservation);
+    }
+
+    // Create reservation event to prevent concurrency issues when deleting
+    function createReservation(err, retrievedDocs, responseOptions) {
+        if (err) throw err;
+        var index = retrievedDocs[0].Index >= 0 ? retrievedDocs[0].Index + 1 : 0;
+        var isAccepted = collection.createDocument(
+            collectionLink,
+            {
+                id: `Event|${aggregateId}[${index}]`,
+                PartitionId: partitionId,
+                AggregateId: aggregateId,
+                Index: index},
+            {},
+            deleteReservation);
+    }
+
+    // Delete reservation
+    function deleteReservation(err, resource, responseOptions) {
+        if (err) throw err;
+        var isAccepted = collection.deleteDocument(resource._self, {}, (err, resource, responseOptions) => {
+            if (err) throw err;
+            // Continue with querying and deleting
+            tryQueryAndDelete();
+        });
+    }
+    
+    // Recursively runs the query w/ support for continuation tokens.
+    // Calls tryDelete(documents) as soon as the query returns documents.
+    function tryQueryAndDelete(continuation) {
```

Comment:
How deep does this recurse? Should you maybe replace recursion with iteration, since it's always faster?

Reply:
I think until some timeout has been reached. Then it returns an object with continuation=true and our code with trigger this procedure again (since there are documents still to delete)

Reply:
The tryQueryAndDelete and tryDelete functions are from their stored procedure examples: https://github.com/Azure/azure-cosmosdb-js-server/blob/master/samples/stored-procedures/bulkDelete.js